sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -373,11 +378,22 @@ public WriterImpl(CassandraIO.Write<T> spec) {
     @Override
     public void write(T entity) {
       Mapper<T> mapper = (Mapper<T>) mappingManager.mapper(entity.getClass());
-      mapper.save(entity);
+      this.writeFutures.add(mapper.saveAsync(entity));
+      if (this.writeFutures.size() % CONCURRENT_ASYNC_QUERIES == 0) {
+        // We reached the max number of allowed in flight queries
+        LOG.debug("Waiting for a batch of 100 Cassandra writes to be executed...");
+        Futures.successfulAsList(this.writeFutures);
```

Comment:
Does it make sense to check that there were some failures (if any) and, perhaps optionally, stop a pipeline with providing some errors/exceptions in logs for user?

Reply:
Yes, we definitely should and successfulAsList() won't do this properly.

I'll loop over the futures instead and call `.get()` so that we can throw the exceptions that would be triggered.
What's the way of dealing with exceptions in Beam IOs ? Should I log the error and re-throw the exception or just not catch it and let it propagate up the stack ?

Reply:
I think, we have to fail the whole bundle in case of failures. I'd recommend to take a look on methods `BigtableWriterFn.checkForFailures()` or `KinesisWriterFn.checkForFailures()`.

Reply:
@aromanenko-dev correct, plus optionally retry write with exponential backoff if that makes sense with the backend